pkg/server/image_pull.go

```diff
@@ -280,68 +262,63 @@ func (c *criService) getTLSConfig(registryTLSConfig config.TLSConfig) (*tls.Conf
 	return tlsConfig, nil
 }
 
-// getResolver tries registry mirrors and the default registry, and returns the resolver and descriptor
-// from the first working registry.
-func (c *criService) getResolver(ctx context.Context, ref string, cred func(string) (string, string, error)) (remotes.Resolver, imagespec.Descriptor, error) {
-	refspec, err := reference.Parse(ref)
-	if err != nil {
-		return nil, imagespec.Descriptor{}, errors.Wrap(err, "parse image reference")
-	}
+// registryHosts is the registry hosts to be used by the resolver.
+func (c *criService) registryHosts(auth *runtime.AuthConfig) docker.RegistryHosts {
```

Comment:
Not sure how it was handled before but it is really better to handle this as a callback. It is best to assume that the credentials may not be accessible until they are requested by the server. Also, these auth configs should only be used when the host matches the server in the auth config. So for example if you are using a mirror of the Docker Hub and have hub credentials, those credentials should not be sent to the mirror, but the mirror can return 404 and have the client retry with the hub and send the credentials. Alternatively, in the configuration you should can explicitly authorize the mirror for another servers credentials if that is a use case you need to support.

Reply:
> Not sure how it was handled before but it is really better to handle this as a callback.

Are you talking about the passed in `AuthConfig`?

> These auth configs should only be used when the host matches the server in the auth config

@dmcgowan The `ServerAddress` in the CRI `AuthConfig` is never properly set... https://github.com/kubernetes/enhancements/pull/1171#discussion_r310348605

It is something we should fix...

Reply:
> Are you talking about the passed in AuthConfig ?

Yes, ideally credentials should not be pre-fetched. Even if they are, the interface shouldn't require it.

> It is something we should fix...

Hmm, I guess the interface should pretend like it is doing the right thing at least, so it is easy to fix later. By "never properly set" do you mean it is empty or set to a completely useless value?

Reply:
@dmcgowan Even if users set `ServerAddress` in their docker file, once imported into Kubernetes, it is lost...

Reply:
So if it is empty, can we use it for all hosts. If it is set, then match the host. In this case, if Kubernetes fixes this issue, then CRI will already handle it properly.

If it is just set to the wrong value, that is harder to use.

Reply:
> So if it is empty, can we use it for all hosts. If it is set, then match the host. In this case, if Kubernetes fixes this issue, then CRI will already handle it properly.

SG. I was just thinking about that. :)

> Yes, ideally credentials should not be pre-fetched. Even if they are, the interface shouldn't require it.

What "credentials" are you referring to? The `tokens` in the `AuthConfig`? Or the `username`/`password`/`auth`?

BTW, can you briefly explain what `RegistryToken` and `IdentityToken` mean in the Docker `AuthConfig`? I believe those are copied from docker api in the first place, and we've never used them in Kubernetes.

Reply:
> So if it is empty, can we use it for all hosts. If it is set, then match the host. In this case, if Kubernetes fixes this issue, then CRI will already handle it properly.

Done